### PR TITLE
Cluster-autoscaler: cloud provider interface

### PR DESCRIPTION
**Pull request by mwielgus**
https://github.com/kubernetes/kubernetes/tree/master/pkg/cloudprovider doesn't define a concept of NodePool/Group/Mig and doesn't have add/delete node in the main interface (although some implementation have it). However we may reuse some fragments of K8s implementation. We could also expand the kubernetes interface although we are not sure how/if the similar capabilities exist in other cloud providers and how to implement them. Once we have some implementations in contrib we may think about moving them to K8s.

Ref: #1311
cc: @piosz @jszczepkowski @fgrzadkowski

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/cloud_provider.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	// NodeGroups returns all node groups configured for this cloud provider.
+	NodeGroups() []*NodeGroup
+
+	// NodeGroupForNode return the node group for the given node, nil if the node
+	// should not be processed by cluster autoscaler, or non-nill error if such
+	// occurred.
+	NodeGroupForNode(*kube_api.Node) (*NodeGroup, error)
+}
+
+// NodeGroup contains configuration info and functionns to control a set
```
Comment (piosz) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: typo functions

File: cluster-autoscaler/cloudprovider/cloud_provider.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+// of nodes that have the same capacity and set of labels.
+type NodeGroup interface {
+	// MaxSize returns maximum size of the node group.
+	MaxSize() int
+
+	// MinSize return minimum size of the node group.
```
Comment (piosz) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: typo returns

File: cluster-autoscaler/cloudprovider/cloud_provider.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	MinSize() int
+
+	// Size returns the current size of the node group.
+	Size() (int, error)
+
+	// GetSampleNode returns a sample node that belongs to this node group.
+	SampleNode() *kube_api.Node
```
Comment (piosz) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: error here?

File: cluster-autoscaler/cloudprovider/cloud_provider.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+	// IncreaseSize increases the size of the node group. To delete a node you need
+	// to explicitely name it and use CloudProvider.
+	IncreaseSize(newNodeCount int) error
+
+	// DeleteNode deletes the node from this node group. Error is returned either on
+	// failue or if the given node doesn't blong to this node group.
```
Comment (piosz) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: typo failure

File: cluster-autoscaler/cloudprovider/cloud_provider.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	// MinSize return minimum size of the node group.
+	MinSize() int
+
+	// Size returns the current size of the node group.
+	Size() (int, error)
+
+	// GetSampleNode returns a sample node that belongs to this node group.
+	SampleNode() *kube_api.Node
+
+	// IncreaseSize increases the size of the node group. To delete a node you need
```
Comment (piosz) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Please describe semantic of the function better.

File: cluster-autoscaler/cloudprovider/cloud_provider.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	// to explicitely name it and use CloudProvider.
+	IncreaseSize(newNodeCount int) error
+
+	// DeleteNode deletes the node from this node group. Error is returned either on
```
Comment (piosz) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Same here

File: cluster-autoscaler/cloudprovider/cloud_provider.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+// CloudProvider contains configuration info and functions for interacting with
+// cloud provider (GCE, AWS, etc).
+type CloudProvider interface {
+	// Name returns name of the cloud provider.
+	Name() string
+
+	// NodeGroups returns all node groups configured for this cloud provider.
+	NodeGroups() []*NodeGroup
+
+	// NodeGroupForNode return the node group for the given node, nil if the node
```
Comment (piosz) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: typo: returns

File: cluster-autoscaler/cloudprovider/cloud_provider.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+	// should not be processed by cluster autoscaler, or non-nill error if such
```
Comment (piosz) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: typo non-nil

File: cluster-autoscaler/cloudprovider/cloud_provider.go (hunk 1 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Done.

File: cluster-autoscaler/cloudprovider/cloud_provider.go (hunk 2 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Done.

File: cluster-autoscaler/cloudprovider/cloud_provider.go (hunk 3 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Done.

File: cluster-autoscaler/cloudprovider/cloud_provider.go (hunk 4 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Done.

File: cluster-autoscaler/cloudprovider/cloud_provider.go (hunk 5 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Done.

File: cluster-autoscaler/cloudprovider/cloud_provider.go (hunk 7 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Done

File: cluster-autoscaler/cloudprovider/cloud_provider.go (hunk 6 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Done.

File: cluster-autoscaler/cloudprovider/cloud_provider.go (hunk 8 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Done.

**Comment by mwielgus** [comment]
cc: @pbitty @iterion

File: cluster-autoscaler/cloudprovider/cloud_provider.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+	Size() (int, error)
+
+	// GetSampleNode returns a sample node that belongs to this node group, or error
+	// if occurred.
+	SampleNode() (*kube_api.Node, error)
+
+	// IncreaseSize increases the size of the node group. To delete a node you need
+	// to explicitely name it and use CloudProvider. This function should wait until
+	// the node appears in the Kubernetes cluster.
+	IncreaseSize(newNodeCount int) error
```
Comment (pbitty) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Do we want the desired total count, or just the delta?

I think we should only need to tell the `NodeGroup` how many instances to add, and not have to be concerned with its total size.  Thoughts?

Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Positive delta. Cluster Autoscaler will not delete nodes without naming them explicitly.

Comment (pbitty) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Makes sense.  I think the docs should explain that `newNodeCount` is the positive delta.

When I first read `newNodeCount` I wasn't sure if it meant the delta or the "new total".

Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Renamed.

File: cluster-autoscaler/cloudprovider/cloud_provider.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+	MinSize() int
+
+	// Size returns the current size of the node group.
+	Size() (int, error)
+
+	// GetSampleNode returns a sample node that belongs to this node group, or error
+	// if occurred.
+	SampleNode() (*kube_api.Node, error)
+
+	// IncreaseSize increases the size of the node group. To delete a node you need
```
Comment (piosz) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: >To delete a node you need to explicitely name it and use CloudProvider.

Is this by mistake?

File: cluster-autoscaler/cloudprovider/cloud_provider.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+	// to explicitely name it and use CloudProvider. This function should wait until
+	// the node appears in the Kubernetes cluster.
+	IncreaseSize(delta int) error
+
+	// DeleteNode deletes the node from this node group. Error is returned either on
+	// failure or if the given node doesn't blong to this node group. This funciton
```
Comment (piosz) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: typo: function

**Comment by fgrzadkowski** [comment]
@mwielgus As discussed offline, in the PR description please explain the motivation for adding this new interface instead of using an existing cloud provider interface that exists in kubernetes.

**Comment by piosz** [comment]
LGTM

**Comment by mwielgus** [comment]
https://github.com/kubernetes/kubernetes/tree/master/pkg/cloudprovider doesn't define a concept of NodePool/Group/Mig and doesn't have add/delete node in the main interface (although some implementation have it). However we may reuse some fragments of K8s implementation. We could also expand the kubernetes interface although we are not sure how/if the similar capabilities exist in other cloud providers and how to implement them. Once we have some implementations in contrib we may think about moving them to K8s.

File: cluster-autoscaler/cloudprovider/cloud_provider.go (hunk 10 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/cloud_provider.go]: Yep, fixed.

**Pull request merged**